docarray/utils/reduce.py

```diff
@@ -0,0 +1,164 @@
+from docarray import DocumentArray
+from typing import List, Optional, Dict, _GenericAlias  # type: ignore
+from docarray.base_document import BaseDocument
+
+
+def _types_analysis(doc: 'BaseDocument') -> List[List[str]]:
+    simple_non_empty_fields: List[str] = []
+    list_fields: List[str] = []
+    set_fields: List[str] = []
+    nested_docs_fields: List[str] = []
+    nested_docarray_fields: List[str] = []
+
+    for field_name, field in doc.__fields__.items():
+        field_type = field.outer_type_
```

Comment:
```suggestion
        field_type = doc._get_field_type(field_name)
```

Reply:
will mypy not complain about accessing a private method?

Reply:
nop I don't think it is an antipattern in python to access private method of other object when you are inside the same library. Python do it all the time itself with all of the `__dunder__` method and attribute

Reply:
no as long as the types match up it will no complain

Reply:
ok